weld/subsystem/src/main/java/org/jboss/as/weld/services/bootstrap/WeldExecutorServices.java

```diff
@@ -115,4 +120,54 @@ public synchronized ExecutorService getTaskExecutor() {
     public void cleanup() {
         // noop on undeploy - the executor is a service shared across multiple deployments
     }
+
+    class WeldExecutor extends ThreadPoolExecutor {
+
+        WeldExecutor(int nThreads, ThreadFactory threadFactory) {
+            super(nThreads, nThreads,
+                               0L, TimeUnit.MILLISECONDS,
+                               new LinkedBlockingQueue<Runnable>(),
+                               threadFactory);
+        }
+
+        @Override
+        protected void beforeExecute(Thread t, Runnable r) {
+            if (r instanceof WeldTaskWrapper) {
+                NamespaceContextSelector.pushCurrentSelector(((WeldTaskWrapper) r).currentSelector);
+            }
+        }
+
+        @Override
+        protected void afterExecute(Runnable r, Throwable t) {
+            if (r instanceof WeldTaskWrapper) {
+                NamespaceContextSelector.popCurrentSelector();
+            }
+        }
+
+        @Override
+        public void execute(Runnable command) {
+            if (Container.instance().getState() == ContainerState.INITIALIZED) {
+                WeldTaskWrapper task = new WeldTaskWrapper(command, NamespaceContextSelector.getCurrentSelector());
+                super.execute(task);
+            } else {
+                super.execute(command);
+            }
+        }
+    }
+
+    class WeldTaskWrapper implements Runnable {
```

Comment:
This class and `WeldExecutor` do not seem to need to access the enclosing class (`WeldExecutorServices`) so it might make sense to make it `static`.

Reply:
@spyrkob Can you make this change? Thanks.

Reply:
Sorry, I missed that. It's updated now.